Better timeouts handling

osipxd opened the pull request:
<!-- 
If this PR updates documentation, please update all relevant versions of the docs, see: https://github.com/kotest/kotest/tree/master/documentation/versioned_docs
The documentation at https://github.com/kotest/kotest/tree/master/documentation/docs is the documentation for the next minor or major version _TO BE RELEASED_
-->
Addressed some issues related to timeouts:
- Use wall clock time for timeouts if `TestCoroutineScheduler` is present in coroutine context (fixes #3703)
- Pass the `timeout` parameter into `runTest` instead of `InvocationTimeoutInterceptor` if `coroutineTestScope` is enabled (fixes #3969)
- Add an original exception as a cause into `TestTimeoutException`

TODO:
- [x] Write tests

osipxd committed:
Refactor invocation timeout handling logic

- Calculate invocation timeout value during config resolution
- Skip InvocationTimeoutInterceptor if invocationTimeout is not configured
- Make timeout and invocationTimeout non-nullable

osipxd committed:
Handle timeouts inside TestCoroutineInterceptor

Pass timeout into runTest if coroutineTestScope is enabled

osipxd committed:
Use wallclock timeout if TestCoroutineScheduler found

osipxd committed:
Add cause to TestTimeoutException

Kantis commented on kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/test/interceptors/BlockedThreadTimeoutInterceptor.kt:
```diff
@@ -84,4 +84,10 @@ internal class BlockedThreadTimeoutInterceptor(
+class BlockedThreadTestTimeoutException(timeout: Duration, testName: String, cause: Throwable? = null) :
+   TestTimeoutException(timeout, testName, cause) {
+
+   @Suppress("unused")
+   @Deprecated("Maintained for binary compatibility", level = DeprecationLevel.HIDDEN)
+   constructor(timeout: Duration, testName: String) : this(timeout, testName, cause = null)
```
💯

Kantis commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TestCoroutineInterceptor.kt:
```diff
@@ -26,11 +27,21 @@ class TestCoroutineInterceptor : TestExecutionInterceptor {
+      // to avoid the influence of default timeout set inside runTest
+      val timeout = testCase.config.invocationTimeout
+      if (timeout != null) {
+         runTest(timeout = timeout, testBody = testBody)
```
Would this apply `invocationTimeout` for `TestType.Container` as well?

osipxd replied:
Oh, I completely forgot to check the test type. Thank you! I've added this check.

Kantis reviewed (commented):
Would appreciate 2nd opinions from @OliverO2 :) 

I think this overall looks good. I'm just wondering about if this breaks things for slow containers which have previously been ignored by `invocationTimeout`? On the other hand, doesn't it make sense to also prevent the test runner from hanging indefinitely if a container gets stuck somehow? 🤔

sksamuel commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/resolveConfig.kt:
```diff
@@ -72,7 +72,11 @@ internal fun resolveConfig(
-      invocationTimeout = invocationTimeout,
+      // note: the invocation timeout cannot be larger than the test case timeout
+      invocationTimeout = when {
+         timeout != null && invocationTimeout != null -> invocationTimeout.coerceAtMost(timeout)
```
Should we just error here so the user is clear it's setup wrong ?

osipxd replied:
Done! But I'm not sure if I throw a correct exception (`IllegalStateException`) and if I do it in the proper place.

sksamuel commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TimeoutInterceptor.kt:
```diff
@@ -32,19 +46,67 @@ internal class TimeoutInterceptor(
+   }
+}
+
+private class KotestTimeoutCancellationException(message: String) : CancellationException(message)
```
could we add a comment on why we've added our own exception type

osipxd replied:
Added 👌

osipxd commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TimeoutInterceptor.kt:
```diff
@@ -32,19 +46,68 @@ internal class TimeoutInterceptor(
-         } catch (t: Throwable) {
-            logger.log { Pair(testCase.name.testName, "Caught timeout $t") }
-            TestResult.Error(mark.elapsedNow(), TestTimeoutException(timeout, testCase.name.testName))
+         } catch (t: CancellationException) {
```
Was there some reason to catch `Throwable` instead of a more specific exception type like `CancellationException` or `TimeoutCancellationException`?

OliverO2 replied:
Catching `Throwable` seems like a mistake to me. Shouldn't we be even more specific, catching just `TimeoutCancellationException` and `KotestTimeoutCancellationException` (which might be named `WallclockTimeoutCancellationException` to better communicate intent)?

OliverO2 replied:
OK, I see that narrowing the Cancellation types has been done inside the catch block.

OliverO2 reviewed (commented):
Overall looks good to me so far.

Still needs some more tests for final review, right?
And wouldn't this also fix #3969?

OliverO2 commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TestCoroutineInterceptor.kt:
```diff
@@ -26,11 +28,21 @@ class TestCoroutineInterceptor : TestExecutionInterceptor {
+         runTest(timeout = timeout, testBody = testBody)
+      } else {
+         runTest(testBody = testBody)
+      }
```
I know that this is unrelated to this PR, but this `runTest` invocation occurs in common code, and its test result is discarded instead of being returned immediately to the encompassing (JS) test framework. [This is described to cause losing test errors.](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-test-result/)

osipxd replied:
As I know, currently, it is impossible to use `TestCoroutinesInterceptor` with JS. It is added only if this condition is true:
```kotlin
val useCoroutineTestScope = platform != Platform.JS && testCase.config.coroutineTestScope
```
So it doesn't affect JS tests yet.

But I agree this issue should be addressed before `coroutineTestScope` is enabled for JS.

OliverO2 replied:
That makes sense. Thanks for pointing it out. Maybe you could grab the opportunity and [add a comment why `coroutineTestScope` is excluded on JS](https://github.com/kotest/kotest/blob/85f6370c608c955030260dd14d393730b3834829/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/TestCaseExecutor.kt#L83).

osipxd replied:
Should I also create an issue to track this change?

OliverO2 replied:
That would be super nice!

osipxd replied:
Here it is: https://github.com/kotest/kotest/issues/4077

osipxd replied:
Added this comment in a separate PR: https://github.com/kotest/kotest/pull/4078
To make this change visible in changelogs.

osipxd committed:
Add tests for timeout with coroutineTestScope enabled

osipxd committed:
Turn ExpectFailureExtension into an object

This change allows ExpectFailureExtension to be specified in particular test config, not only for the whole spec.

osipxd commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TimeoutInterceptor.kt:
```diff
@@ -32,19 +46,68 @@ internal class TimeoutInterceptor(
+      timeoutJob.onJoin {
+         blockDeferred.cancel()
+         throw WallclockTimeoutCancellationException("Timed out waiting for $timeout")
+      }
```
It doesn't work 🤔
The exception seems to be ignored. I've tried several other workarounds with no luck. So I need some time to investigate the reason.

osipxd reviewed (commented):
I've added tests.
Force-pushed the branch after it, so here is the [link to the correct diff](https://github.com/kotest/kotest/compare/64e26c0f80e3516e9908d869d61c627ce5e7d0de..4fafbe454e174654f93a40e6cb92cc71ea97557f)

osipxd committed:
Run timeout tests for both values of coroutineTestScope

osipxd commented on kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/test/config/ResolvedTestConfig.kt:
```diff
@@ -30,13 +32,13 @@ data class ResolvedTestConfig(
     * has the same effect as timeout. To set a timeout across all invocations then see [timeout].
     */
-   val invocationTimeout: Duration?,
+   val invocationTimeout: Duration,
```
As I can see it is safe to make these fields non-nullable. The only place where these fields were `null` was `ResolvedTestConfig.default`, used only for Kotest tests and always replaced by a resolved configuration in real test execution.

osipxd replied:
The only problem is the [binary compatibility](https://jakewharton.com/public-api-challenges-in-kotlin/#binary-compatibility) of the data classes. I can provide a secondary constructor, copy method and destructuring operators with old signatures to maintain binary compatibility, but it is a pain.

AlexCue987 commented on kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/test/config/ResolvedTestConfig.kt:
```diff
@@ -92,8 +97,8 @@ data class ResolvedTestConfig(
-         timeout = null,
-         invocationTimeout = null,
+         timeout = Defaults.defaultTimeoutMillis.milliseconds,
+         invocationTimeout = Defaults.defaultInvocationTimeoutMillis.milliseconds,
```
the following line says "miliseconds" twice: `Defaults.defaultInvocationTimeoutMillis.milliseconds` - is that redundant?

osipxd replied:
Yes, `Defaults.defaultInvocationTimeoutMillis` is `Long`, while I need `Duration` here.

osipxd commented on kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/test/config/ResolvedTestConfig.kt:
```diff
@@ -81,6 +83,9 @@ data class ResolvedTestConfig(
    }
 
+   // note: the invocation timeout can't be larger than the test case timeout
+   val resolvedInvocationTimeout: Duration = minOf(timeout, invocationTimeout)
```
Feel free to propose a better naming for this field. Maybe, it should be internal.

osipxd replied:
Removed this field

osipxd commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/ContainerTimeoutTest.kt:
```diff
@@ -1,38 +1,42 @@
-               .async()
-         }
-
-         duration shouldBe 100.milliseconds
```
After removing `coroutineTestScope` from the Spec itself, duration is not deterministic anymore, that's why I decided to remove this assertion.

AlexCue987 commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TestCoroutineInterceptor.kt:
```diff
@@ -28,7 +29,15 @@ class TestCoroutineInterceptor : TestExecutionInterceptor {
-      runTest {
+
+      // Handle timeouts here to avoid the influence of the default timeout set inside runTest
+      val timeout = if (testCase.type == TestType.Container) {
```
should we delegate this logic to `testCase` so that we avoid its tight coupling with this interseptor?

AlexCue987 commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TimeoutInterceptor.kt:
```diff
@@ -23,28 +38,79 @@ internal class TimeoutInterceptor(
-            }
-         } catch (t: Throwable) {
+      // each test takes 300ms, and a timeout of 800ms, this would fail, because 3 x 300 > 800.
+      logger.log { Pair(testCase.name.testName, "Switching context to add timeout $timeout") }
```
Nit: do we need a `Pair` to construct a text message to log?

osipxd replied:
`Logger.log` takes `() -> Pair<String?, String>` as a parameter, so I think yes

AlexCue987 commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/CoroutinesTimeoutTest.kt:
```diff
@@ -88,3 +95,20 @@ private suspend fun someCoroutine() {
 }
+
+/** Delay using real (non-virtual) time. */
+suspend fun wallclockDelay(duration: Duration) {
```
should we rename this function to something like `nonVirtualTimeDelay` to design away the need for this comment?

AlexCue987 commented on kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TimeoutInterceptor.kt:
```diff
@@ -23,28 +38,79 @@ internal class TimeoutInterceptor(
+   }
+}
+
+private suspend fun <T> withWallclockTimeout(
```
this function is quite large. does it have its own unit tests?

osipxd replied:
It is tricky to test it. I have only a test `ContainerTimeoutOverriddenMainDispatcherTest` failing without this function, but I'll try to write a separate test.

AlexCue987 reviewed (commented):
LGTM, some optional cosmetic suggestions

osipxd committed:
Move timeout logic to extension-function on TestCase

osipxd committed:
Rename wallclock to nonVirtualTime

AlexCue987 committed:
Merge branch 'master' into feature/better-timeouts

OliverO2 commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/ContainerTimeoutOverriddenMainDispatcherTest.kt:
```diff
@@ -0,0 +1,45 @@
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.minutes
+
+// Issue: https://github.com/kotest/kotest/issues/3703
+@OptIn(ExperimentalCoroutinesApi::class)
+class ContainerTimeoutOverriddenMainDispatcherTest : FunSpec({
```
```suggestion
// Issue: https://github.com/kotest/kotest/issues/3703
@DoNotParallelize
@OptIn(ExperimentalCoroutinesApi::class)
class ContainerTimeoutOverriddenMainDispatcherTest : FunSpec({
```

Using/changing `Dispatchers.Main` depends on a global object. This does not play well with any type of concurrency. `@DoNotParallelize` will run this test in isolation, so that it cannot interfere with possible tests elsewhere which also happen to use/change `Dispatchers.Main`.

OliverO2 commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/CoroutinesTimeoutOverrideTest.kt:
```diff
@@ -0,0 +1,35 @@
+import io.kotest.core.spec.style.FunSpec
+import kotlin.time.Duration.Companion.milliseconds
+
+class CoroutinesTimeoutOverrideTest : FunSpec({
```
```suggestion
@DoNotParallelize
class CoroutinesTimeoutOverrideTest : FunSpec({
```

`System.setProperty` affects global state.

OliverO2 commented on kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/timeout/CoroutinesTimeoutTest.kt:
```diff
@@ -88,3 +95,20 @@ private suspend fun someCoroutine() {
    }
 }
+
+suspend fun nonVirtualTimeDelay(duration: Duration) {
```
At your discretion: Would `realTimeDelay` be a better fit for our human wetware, avoiding negativity?

OliverO2 reviewed (commented):
Thanks for the huge effort!

Some general aspects:
* Mixing virtual and real time (timeouts) is hard to reason about. I've tried to spot potential issues, but I'm not sure that I have a full understanding of all the implications.
* Normally, I would not expect a test running on virtual time to require any real-time timeout, as such tests typically complete quickly.
* Some real-time timeouts introduced or changed in this PR may result in flaky tests, in particular on our slow CI runners (macOS, Windows). We have switched most time-dependent tests to virtual time to avoid that. To stabilize tests with real-time timeouts may require large time intervals (seconds), which slows down tests. See https://github.com/kotest/kotest/issues/4113.

Hope this helps!
